### PR TITLE
Deprecate UnrolledBuffer.waterlineDelim

### PR DESCRIPTION
**Pull request by joshlemer**

### CODE REVIEW

**Review by NthPortal — COMMENTED** [review]
LGTM. @julienrf?

**Review by julienrf — APPROVED** [review]

**Comment by NthPortal** [comment]
It looks like `SerializationStabilityTest` needs to be updated.

If you look at the top of `SerializationStabilityTest.scala`, you will find the following comment

```
// This test is self-modifying when run as follows:
//
//    junit/testOnly scala.SerializationStabilityTest -- -Doverwrite.source=test/junit/scala/SerializationStabilityTest.scala
//
// Use this to re-establish a baseline for serialization compatibility.
```

Running the specified sbt command will update `SerializationStabilityTest` (to take into account the structural changes to classes/traits in your PR which cause serialization to change). After running the sbt command, `SerializationStabilityTest` will pass.

Modifying `SerializationStabilityTest` is okay before a major release (e.g. before 2.13.0), because there is no prior version with which serialization compatibility is required. However, modifying `SerializationStabilityTest` is not okay before a minor release (e.g. after 2.13.0 but before 2.13.1), because the new minor release must have its serialization be compatible with the previous minor version.

**Comment by szeiger** [comment]
> owever, modifying SerializationStabilityTest is not okay before a minor release (e.g. after 2.13.0 but before 2.13.1), because the new minor release must have its serialization be compatible with the previous minor version.

AFAIK we do not guarantee any serialization stability even between minor releases.

File: src/library/scala/collection/mutable/UnrolledBuffer.scala [hunk 1]
```diff
@@ -237,8 +237,13 @@ object UnrolledBuffer extends StrictOptimizedClassTagSeqFactory[UnrolledBuffer]
...
   def newBuilder[A : ClassTag]: UnrolledBuffer[A] = new UnrolledBuffer[A]
 
-  val waterline = 50
-  val waterlineDelim = 100    // TODO -- fix this name!  It's a denominator, not a delimiter.  (But it's part of the API so we can't just change it.)
+  final val waterline: Int = 50
+
+  final val waterlineDenom: Int = 100
+
+  @deprecated("Use waterlineDenom instead.", "2.13.0")
+  final val waterlineDelim: Int = waterlineDenom
```
Comment (szeiger) [inline src/library/scala/collection/mutable/UnrolledBuffer.scala:249]: Is there any reason why this has to be a `val`? A `def` doesn't require an extra field and keeps serialization stability.

Comment (joshlemer) [reply src/library/scala/collection/mutable/UnrolledBuffer.scala:249]: Not that I'm aware of, I can change that 👍

**Comment by NthPortal** [comment]
@szeiger do we not? I'll update my saved reply to say to update the `@SerialVersionUID` instead then

**Comment by lrytz** [comment]
I'm not sure. @retronym? Let's add this to https://docs.scala-lang.org/overviews/core/binary-compatibility-of-scala-releases.html

**Comment by adriaanm** [comment]
We try to maintain serialization compatibility, but since we don't check it mechanically, we can't really pinky-swear-promise it.

**Comment by NthPortal** [comment]
@adriaanm we do have `SerializationStabilityTest` now though. It doesn't check everything, but it catches a lot

**Comment by adriaanm** [comment]
Good point, but it's not exhaustive in the same way MiMa is.

**Review by szeiger — APPROVED** [review]
LGTM. Needs to be rebased.

**Comment by NthPortal** [comment]
@adriaanm @szeiger should we consider `SerializationStabilityTest` frozen within a major release? Or mostly frozen? Under what conditions should it be changed, and under what conditions not?

**Comment by szeiger** [comment]
We haven't guaranteed it so far and it does not appear to be a big problem in practice. I'd be hesitant to add any guarantees around serialization compatibility because it will make it even harder to change anything in a compatible way in minor releases. Plus we'd first need tooling to check for incompatible changes, similar to what MiMa does for binary compatibility.

**Comment by SethTisue** [comment]
the lack of any documentation (that I know of, anyway?) on how much serialization stability we do or don't guarantee now has a ticket over at https://github.com/scala/docs.scala-lang/issues/1202

**Comment by NthPortal** [comment]
I'm sorry for hijacking this PR by using a saved reply that wasn't quite accurate 😂

**Comment by szeiger** [comment]
Anyway, this is for 2.13.x, so serialization stability is of no concern. This just needs to be rebased.

**Comment by adriaanm** [comment]
Rebased. I ran ` junit/testOnly scala.SerializationStabilityTest ` after a bootstrap locally, I wonder why no changes were required on top of current 2.13.x.

**Comment by joshlemer** [comment]
❗️ Remember not to merge until as @szeiger requested, the `waterlineDenom` and `waterlineDelim` are `def`s not `val`s

**Comment by adriaanm** [comment]
Thanks for the reminder! I had missed that.

**Comment by SethTisue** [comment]
@joshlemer interested in returning to this now...?

File: src/library/scala/collection/mutable/UnrolledBuffer.scala (hunk 1 above)
Comment (joshlemer) [reply src/library/scala/collection/mutable/UnrolledBuffer.scala:249]: @szeiger I am assuming you meant to keep waterline and waterlineDelim as a `val` but set waterlineDenom as a def, that's what I've done now.

**Comment by joshlemer** [comment]
@SethTisue Change have been pushed, waiting on @szeiger to confirm I interpreted his comment correctly.

File: src/library/scala/collection/mutable/UnrolledBuffer.scala (hunk 1 above)
Comment (Ichoran) [reply src/library/scala/collection/mutable/UnrolledBuffer.scala:249]: I think breakage is less likely if the def references the val, not the other way around.  On the other hand, is it a problem to make everything a `def`?  There's no reason any of this should be serialized.

**Comment by szeiger** [comment]
The original discussion here predates https://github.com/scala/scala/pull/7297 which eliminated serialization of object state. This is the reason why it doesn't affect serialization stability anymore. I'd make both `waterline` and `waterlineDenom` a `def` for consistency and keep the deprecated `waterlineDelim` as a `val` but I'm OK merging it as it is. This bike shed has been repainted too many times already.

**Comment by SethTisue** [comment]
> This bike shed has been repainted too many times already

haha yeah, I mentally groaned when I saw the ticket name pop up in my notifications 😜

regardless, Josh, we are grateful for your attention to this

**Comment by SethTisue** [comment]
@joshlemer feel free to either revise as Stefan suggests and then merge once CI likes it, or just merge as is, as you see fit.

**Pull request merged**